lib/rust/parser/src/syntax/operator.rs

```diff
@@ -0,0 +1,171 @@
+/// not needed. For example, `(.x)` is parsed as `(\t -> t.x)`, which is understandable.
+fn annotate_tokens_that_need_spacing(items: Vec<syntax::Item>) -> Vec<syntax::Item> {
+    // TODO: It should be possible to make it faster by iterating over mut vec. To be checked.
+    items
+        .into_iter()
+        .map(|item| match item {
+            syntax::Item::Block(_) => item,
+            syntax::Item::Token(_) => item,
+            syntax::Item::Tree(ast) => syntax::Item::Tree(match &*ast.variant {
+                syntax::tree::Variant::MultiSegmentApp(data)
+                    if !data.segments.first().header.is_symbol() =>
+                    ast.with_error("This expression cannot be used in a non-spaced equation."),
+                _ => ast,
+            }),
+        })
+        .collect()
+}
+
+/// Take [`Item`] stream, resolve operators precedence and return the final AST. The precedence
+/// resolution algorithm bases on the [Shunting yard algorithm](https://en.wikipedia.org/wiki/Shunting_yard_algorithm).
+/// It is extended to handle operator sections.
+#[inline(always)]
+pub fn resolve_operator_precedence<'s>(items: Vec<syntax::Item<'s>>) -> syntax::Tree<'s> {
+    type Tokens<'s> = Vec<syntax::Item<'s>>;
+    let mut flattened: Tokens<'s> = default();
+    let mut no_space_group: Tokens<'s> = default();
+    let processs_no_space_group = |flattened: &mut Tokens<'s>, no_space_group: &mut Tokens<'s>| {
+        let tokens = mem::take(no_space_group);
+        if tokens.len() == 1 {
+            flattened.extend(tokens);
+        } else {
+            let tokens = annotate_tokens_that_need_spacing(tokens);
+            let ast = resolve_operator_precedence_internal(tokens);
+            flattened.push(ast.into());
+        }
+    };
+    for item in items {
+        if item.left_visible_offset().width_in_spaces == 0 || no_space_group.is_empty() {
+            no_space_group.push(item)
+        } else if !no_space_group.is_empty() {
+            processs_no_space_group(&mut flattened, &mut no_space_group);
+            no_space_group.push(item);
+        } else {
+            // FIXME: this is unreachable.
+            flattened.push(item);
+        }
+    }
+    if !no_space_group.is_empty() {
+        processs_no_space_group(&mut flattened, &mut no_space_group);
+    }
+    resolve_operator_precedence_internal(flattened)
+}
+
+fn resolve_operator_precedence_internal(items: Vec<syntax::Item<'_>>) -> syntax::Tree<'_> {
+    // Reverse-polish notation encoding.
+    let mut was_section_used = false;
+    let mut output: Vec<syntax::Item> = default();
+    let mut operator_stack: Vec<WithPrecedence<syntax::tree::OperatorOrError>> = default();
+    let mut last_token_was_ast = false;
+    let mut last_token_was_opr = false;
+    for item in items {
+        let i2 = item.clone(); // FIXME: Why is this even needed? Rust bug?
+        if let syntax::Item::Token(token) = i2 && let token::Variant::Operator(opr) = token.variant {
```

Comment:
The `if let` has to consume its rhs in order for the `&& let` pattern to be tested, so the rhs value is not available in the `else` block.

Reply:
Ahhhh, that's right!